redis_commands.c

```diff
@@ -2670,9 +2670,83 @@ int redis_smove_cmd(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
     return SUCCESS;
 }
 
-/* Generic command construction for HSET and HSETNX */
-static int gen_hset_cmd(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
-                        char *kw, char **cmd, int *cmd_len, short *slot)
+/* HSET */
+int redis_hset_cmd(INTERNAL_FUNCTION_PARAMETERS, RedisSock *redis_sock,
+                   char **cmd, int *cmd_len, short *slot, void **ctx)
+{
+    int i, argc;
+    smart_string cmdstr = {0};
+    zend_string *zkey;
+    zval *z_args, *z_ele;
+
+    if ((argc = ZEND_NUM_ARGS()) < 2) {
+        return FAILURE;
+    }
+
+    z_args = ecalloc(argc, sizeof(*z_args));
+    if (zend_get_parameters_array(ht, argc, z_args) == FAILURE) {
+        efree(z_args);
+        return FAILURE;
+    }
+
+    if (argc == 2) {
+        if (Z_TYPE(z_args[1]) != IS_ARRAY || zend_hash_num_elements(Z_ARRVAL(z_args[1])) % 2) {
```

Comment:
Do we need the `zend_hash_num_elements(Z_ARRVAL(z_args[1])) % 2` check, since we're iterating over key => value pairs below?

e.g.

```php
$redis->hSet('somehash', ['one' => 'v', 'two' => 'v', 'three' => 'v']);
```